numpydoc/numpydoc.py

```diff
@@ -203,12 +203,27 @@ def mangle_signature(app, what, name, obj, options, sig, retann):
     if not hasattr(obj, '__doc__'):
         return
     doc = get_doc_object(obj, config={'show_class_members': False})
-    sig = doc['Signature'] or getattr(obj, '__text_signature__', None)
+    sig = (doc['Signature']
+           or _clean_text_signature(getattr(obj, '__text_signature__', None)))
     if sig:
         sig = re.sub("^[^(]*", "", sig)
         return sig, ''
 
 
+def _clean_text_signature(sig):
+    if sig is None:
+        return None
+    start_pattern = re.compile(r"^[^(]*\(")
+    start, end = start_pattern.search(sig).span()
+    start_sig = sig[start:end]
+    sig = sig[end:-1]
+    params = sig.split(', ')
+    for param in ('$self', '$module', '$type', '/'):
+        if param in params:
```

Comment:
I don't think it'll happen often, but I'm still not comfortable with the fact that this processes all parts of the sig. Can't we check explicitly that the `/` precedes a `*`, and that the `$` is at the beginning, rather than use the split and join logic?
```
sig = re.sub(r'^\$(self|module|type), ', '', sig, count=1)
sig = re.sub(r', /, \*', ', *', sig, count=1)
```
Is this not sufficient?

Reply:
I'll confess that I have (almost) no understanding of how these C text signature things work. 

What you have would make it so `func($self)` would still be `func($self)`. Is that alright? I have no idea what the possible syntax is for this.

If you (or someone else) has some kind of documentation on what is possible, that would make this back-and-forth much easier.

Reply:
Sorry, I meant to apply those substitutions after
```
    start_pattern = re.compile(r"^[^(]*\(")
    start, end = start_pattern.search(sig).span()
    start_sig = sig[start:end]
    sig = sig[end:-1]
```
They're just a bit more careful than splitting on every comma.

Reply:
Actually I think this makes sense. Let me change that.

To handle `func($self)`, I guess you could just add a `’, ‘` right before your code. Then use regex to strip off a trailing `’, ‘` if it exists. 

Do you think that would work?

Reply:
Sorry, didn't handle that case, did I?
Can use `r'^\$(self|module|type)(, |$))?'` which will strip a comma, or check that we're at the end.

Reply:
Take a peek at this. I had to do something to check whether the `/` was the start of the string or preceded by a `, `.